src/Utilities/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataSymbolMap.cs

```diff
@@ -0,0 +1,131 @@
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow;
+
+namespace Analyzer.Utilities.FlowAnalysis.Analysis.TaintedDataAnalysis
+{
+    /// <summary>
+    /// Mapping of <see cref="ITypeSymbol"/> to <see cref="ITaintedDataInfo"/> (tainted data source/sanitizer/sink info).
+    /// </summary>
+#pragma warning disable CA1812   // TaintedDataSymbolMap is used in other projects.
+    internal class TaintedDataSymbolMap<TInfo> : IEquatable<TaintedDataSymbolMap<TInfo>>
+        where TInfo : ITaintedDataInfo
+#pragma warning restore CA1812
+    {
+        public TaintedDataSymbolMap(WellKnownTypeProvider wellKnownTypeProvider, IEnumerable<TInfo> taintedDataInfos)
+        {
+            if (wellKnownTypeProvider == null)
+            {
+                throw new ArgumentNullException(nameof(wellKnownTypeProvider));
+            }
+
+            if (taintedDataInfos == null)
+            {
+                throw new ArgumentNullException(nameof(taintedDataInfos));
+            }
+
+            ImmutableDictionary<ITypeSymbol, TInfo>.Builder concreteInfosBuilder = ImmutableDictionary.CreateBuilder<ITypeSymbol, TInfo>();
+            ImmutableDictionary<ITypeSymbol, TInfo>.Builder interfaceInfosBuilder = ImmutableDictionary.CreateBuilder<ITypeSymbol, TInfo>();
+
+            foreach (TInfo info in taintedDataInfos)
+            {
+                if (wellKnownTypeProvider.TryGetTypeByMetadataName(info.FullTypeName, out INamedTypeSymbol namedTypeSymbol))
+                {
+                    if (info.IsInterface)
+                    {
+                        interfaceInfosBuilder[namedTypeSymbol] = info;
+                    }
+                    else
+                    {
+                        concreteInfosBuilder[namedTypeSymbol] = info;
+                    }
+                }
+            }
+
+            this.ConcreteInfos = concreteInfosBuilder.ToImmutable();
+            this.InterfaceInfos = interfaceInfosBuilder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Mapping for concrete types.
+        /// </summary>
+        private ImmutableDictionary<ITypeSymbol, TInfo> ConcreteInfos { get; }
+
+        /// <summary>
+        /// Mapping for interface types.
+        /// </summary>
+        private ImmutableDictionary<ITypeSymbol, TInfo> InterfaceInfos { get; }
+
+        /// <summary>
+        /// Indicates that this mapping is empty, i.e. there are no types referenced by the compilation represented by the <see cref="WellKnownTypeProvider"/>.
+        /// </summary>
+        public bool IsEmpty { get { return this.ConcreteInfos.IsEmpty && this.InterfaceInfos.IsEmpty; } }
+
+        /// <summary>
+        /// Gets an enumeration of infos for the given type.
+        /// </summary>
+        /// <param name="namedTypeSymbol">Type to find infos for.</param>
+        /// <returns>Relevant infos for the given type.</returns>
+        public IEnumerable<TInfo> GetInfosForType(INamedTypeSymbol namedTypeSymbol)
```

Comment:
>GetInfosForType [](start = 34, length = 15)

Seems like this method will be invoked for every single property/method reference in compilation. Consider caching here ` ConcurrentDictionary<INamedTypeSymbol, ImmutableHashSet<TInfo>>`. We do something similar for dispose rules: [https://github.com/dotnet/roslyn-analyzers/blob/features/dataflow/src/Utilities/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysisHelper.cs#L152](https://github.com/dotnet/roslyn-analyzers/blob/features/dataflow/src/Utilities/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysisHelper.cs#L152)

Reply:
Also seems like the only callsites for these are iterating over the infos and checking if a given method/property symbol belongs in the TaintedSet within infos. More optimal implementation might just be `ConcurrentDictionary<INamedTypeSymbol, ImmutableHashSet<ISymbol>>, and change this method to return set of tainted symbols, so the callers can just do a single `.Contains` check instead of iterate over all infos each time.

---
In reply to: [234009151](https://github.com/dotnet/roslyn-analyzers/pull/1882#discussion_r234009151) [](ancestors = 234009151)

Reply:
Tracked in https://github.com/dotnet/roslyn-analyzers/issues/1895